### PR TITLE
*: use ring buffer for data send.

### PR DESCRIPTION
**Pull request by siddontang**
@ngaut @BusyJay @disksing @zhangjinpeng1987

### CODE REVIEW

File: src/server/conn.rs [hunk 1]
```diff
@@ -298,7 +298,9 @@ impl Conn {
         // We must also check `socket is not connected` error too, when we connect to a remote
         // store, mio puts this socket in event loop immediately, but this socket may not be
         // connected at that time, so we must register writable too for this case.
-        self.res.push_back(msg.encode_to_buf());
+        let mut send_buffer = self.send_buffer.take().unwrap();
```
Comment (ngaut) [inline src/server/conn.rs]: msg.encode_to(self.send_buffer.as_mut().unwrap()).unwrap()

File: src/server/conn.rs [hunk 2]
```diff
@@ -95,10 +94,13 @@ impl Conn {
...
             file_size: 0,
             payload: None,
-            res: VecDeque::new(),
             last_msg_id: 0,
             snap_scheduler: snap_scheduler,
             store_id: store_id,
+            // send buffer can be grown automatically, first using
+            // DEFAULT_SEND_BUFFER_SIZE is ok. Maybe we should need
+            // max size to shrink later.
+            send_buffer: Some(SendBuffer::new(DEFAULT_SEND_BUFFER_SIZE)),
```
Comment (ngaut) [inline src/server/conn.rs]: Can we remove the option ?

Comment (siddontang) [inline src/server/conn.rs]: No, in write_buf, we will use self.sock and self.send_buffer together, so must use option to take the ownership first.

Comment (BusyJay) [inline src/server/conn.rs]: I don't think taking ownership is necessary because `sock` and `send_buffer` are different fields of `Conn`.

Comment (siddontang) [inline src/server/conn.rs]: oh, yep, my fault.

File: src/util/buf.rs [hunk 3]
```diff
@@ -20,3 +23,82 @@ pub fn create_mem_buf(s: usize) -> MutByteBuf {
...
+    }
+
+    pub fn send<T: Write>(&mut self, w: &mut T) -> Result<usize> {
+        let count = {
+            let (left, right) = self.buf.as_slices();
+            let res = try!(w.try_write(left));
+            let mut count = 0;
+            match res {
+                None => return Ok(0),
+                Some(n) => count += n,
```
Comment (BusyJay) [inline src/util/buf.rs]: What if `n < left.len()`?

Comment (siddontang) [inline src/util/buf.rs]: oh, should not send right, I will fix it.

File: src/util/buf.rs [hunk 4]
```diff
@@ -36,16 +36,15 @@ impl SendBuffer {
     pub fn send<T: Write>(&mut self, w: &mut T) -> Result<usize> {
         let count = {
             let (left, right) = self.buf.as_slices();
-            let res = try!(w.try_write(left));
-            let mut count = 0;
-            match res {
+            let mut count = match try!(w.try_write(left)) {
```
Comment (BusyJay) [inline src/util/buf.rs]: What if `left.len() == 0`?

Comment (siddontang) [inline src/util/buf.rs]: left is empty, return 0.

Comment (BusyJay) [inline src/util/buf.rs]: Is it possible that left is empty but right is not?

Comment (siddontang) [inline src/util/buf.rs]: yes

File: src/server/conn.rs [hunk 5]
```diff
@@ -253,28 +254,15 @@ impl Conn {
...
-        while !self.res.is_empty() {
-            let remaining = try!(self.write_buf());
-
-            if remaining > 0 {
-                // we don't write all data, so must try later.
-                // we have already registered writable, no need registering again.
-                return Ok(());
-            }
-            self.res.pop_front();
+        try!(self.send_buffer.send(&mut self.sock));
```
Comment (zhangjinpeng1987) [inline src/server/conn.rs]: send_to maybe more clearly

**Comment by siddontang** [comment]
PTAL

**Comment by ngaut** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @BusyJay @zhangjinpeng1987

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**